File: gatling-java/src/test/java/sentinelview/StressTestSimulation.java
package sentinelview;

import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;

import java.time.Duration;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.mqtt.MqttDsl.*;
import static sentinelview.SentinelViewProtocol.*;

/**
 * Staircase stress test — progressively doubles concurrent connections to find the breaking point.
 *
 * Each stage ramps in new users while all previous users stay connected and keep publishing.
 * The duration is split into 5 equal stages:
 *
 *   Stage 1: +users×1  → total: users×1
 *   Stage 2: +users×1  → total: users×2
 *   Stage 3: +users×2  → total: users×4
 *   Stage 4: +users×4  → total: users×8
 *   Stage 5: +users×8  → total: users×16
 *
 * Example with users=50, duration=300 (60s per stage):
 *   50 → 100 → 200 → 400 → 800 concurrent cameras at peak
 *
 * Example with users=200, duration=300:
 *   200 → 400 → 800 → 1600 → 3200 concurrent cameras at peak
 *
 * Look for the inflection point in Gatling Enterprise reports where response times
 * spike or error rates climb — that's the broker's capacity limit.
 *
 * All users loop forever (publish telemetry + alert every 5s) until the test ends.
 * maxDuration ensures a clean stop.
 */
public class StressTestSimulation extends Simulation {

    ScenarioBuilder scn = scenario("Stress Test")
            .feed(cameraFeeder())
            .exec(mqtt("connect").connect())
            .forever().on(                                         // loop until maxDuration kills the test
                    pace(Duration.ofSeconds(5))
                            .exec(session -> {
                                String siteId = session.getString("siteId");
                                String cameraId = session.getString("cameraId");
                                return session
                                        .set("telemetryPayload", telemetryPayload(siteId, cameraId))
                                        .set("alertPayload", alertPayload(siteId, cameraId));
                            })
                            .exec(mqtt("publish telemetry")
                                    .publish("sentinelview/#{siteId}/#{cameraId}/telemetry")
                                    .message(StringBody("#{telemetryPayload}"))
                                    .qosAtMostOnce())
                            .exec(mqtt("publish alert")
                                    .publish("sentinelview/#{siteId}/#{cameraId}/alert")
                                    .message(StringBody("#{alertPayload}"))
                                    .qosAtLeastOnce())
            );

    /** Each stage lasts 1/5th of the total duration */
    Duration stageDuration = Duration.ofSeconds(DURATION_SEC / 5);

    {
        // Staircase injection: each stage adds more users, doubling cumulative total
        setUp(
                scn.injectOpen(
                        rampUsers(USERS).during(stageDuration),          // Stage 1: +1x
                        rampUsers(USERS).during(stageDuration),          // Stage 2: +1x (total 2x)
                        rampUsers(USERS * 2).during(stageDuration),      // Stage 3: +2x (total 4x)
                        rampUsers(USERS * 4).during(stageDuration),      // Stage 4: +4x (total 8x)
                        rampUsers(USERS * 8).during(stageDuration)       // Stage 5: +8x (total 16x)
                )
        ).protocols(MQTT_PROTOCOL)
                .maxDuration(Duration.ofSeconds(DURATION_SEC));          // hard stop
    }
}

File: gatling/src/test/scala/sentinelview/StressTestSimulation.scala
package sentinelview

import scala.concurrent.duration._
import io.gatling.core.Predef._
import io.gatling.mqtt.Predef._

class StressTestSimulation extends Simulation {

  import SentinelViewProtocol._

  // Each user connects, publishes telemetry in a loop until the global maxDuration kills the test.
  // The injection profile ramps in stages so we can observe when the broker starts to degrade.
  val scn = scenario("Stress Test")
    .feed(cameraFeeder)
    .exec(mqtt("connect").connect)
    .forever {
      pace(5.seconds)
        .exec { session =>
          val siteId   = session("siteId").as[String]
          val cameraId = session("cameraId").as[String]
          session
            .set("telemetryPayload", telemetryPayload(siteId, cameraId))
            .set("alertPayload", alertPayload(siteId, cameraId))
        }
        .exec(
          mqtt("publish telemetry")
            .publish("sentinelview/#{siteId}/#{cameraId}/telemetry")
            .message(StringBody("#{telemetryPayload}"))
            .qosAtMostOnce
        )
        .exec(
          mqtt("publish alert")
            .publish("sentinelview/#{siteId}/#{cameraId}/alert")
            .message(StringBody("#{alertPayload}"))
            .qosAtLeastOnce
        )
    }

  // Staircase injection: ramp to each stage over 20s, hold for 40s, then next stage.
  // Stages: users*1, users*2, users*4, users*8, users*16
  // With default users=500: 500 → 1000 → 2000 → 4000 → 8000
  // With users=50 (for testing): 50 → 100 → 200 → 400 → 800
  val stageDuration = (durationSec / 5).seconds

  setUp(
    scn.inject(
      rampUsers(users).during(stageDuration),
      rampUsers(users).during(stageDuration),
      rampUsers(users * 2).during(stageDuration),
      rampUsers(users * 4).during(stageDuration),
      rampUsers(users * 8).during(stageDuration)
    )
  ).protocols(mqttProtocol)
    .maxDuration(durationSec.seconds)
}
